File: aoc24/python/day04.py
with open("aoc24/inputs/day04.txt") as f:
    grid = f.read().splitlines()


def char_at(x, y):
    if x < 0 or y < 0:
        return "."
    try:
        return grid[x][y]
    except IndexError:
        return "."


def read_in_direction(x, y, dx, dy):
    return "".join((char_at(x + n * dx, y + n * dy) for n in range(4)))


def xmas_count(x, y):
    if char_at(x, y) != "X":
        return 0

    directions = [
        [dx, dy] for dx in (-1, 0, 1) for dy in (-1, 0, 1) if dx != 0 or dy != 0
    ]

    return len(
        [
            direction
            for direction in directions
            if read_in_direction(x, y, *direction) == "XMAS"
        ]
    )


def read_cross(x, y):
    return {
        char_at(x - 1, y - 1) + char_at(x, y) + char_at(x + 1, y + 1),
        char_at(x + 1, y - 1) + char_at(x, y) + char_at(x - 1, y + 1),
    }


nrows = len(grid)
ncols = len(grid[0])

part_1 = part_2 = 0
for x in range(nrows):
    for y in range(ncols):
        part_1 += xmas_count(x, y)
        if read_cross(x, y).issubset({"MAS", "SAM"}):
            part_2 += 1

print(f"Part 1: {part_1}")
print(f"Part 2: {part_2}")

File: aoc24/src/day04.clj
(ns aoc24.src.day04
  (:require
   [aoc-tools :refer [read-input submit-answer]]
   [utils :refer [non-neg?]]))

(def input (read-input :test))
(def real-input (read-input))

(defn valid-position?
  [pos input]
  (and
   (< (first pos) (count input))
   (< (second pos) (count input))
   (non-neg? (first pos))
   (non-neg? (second pos))))

(defn char-at
  [pos input]
  (if (valid-position? pos input)
    (nth (nth input (first pos)) (second pos))
    \.))


(defn xmas-in-direction?
  [pos-1 direction input]
  (let [pos-2 (map + pos-1 direction)
        pos-3 (map + pos-2 direction)
        pos-4 (map + pos-3 direction)]
    (and
     (= (char-at pos-1 input) \X)
     (= (char-at pos-2 input) \M)
     (= (char-at pos-3 input) \A)
     (= (char-at pos-4 input) \S))))

(defn xmas-count
  [pos input]
  (let [directions [[0 1] [0 -1] [1 0] [-1 0] [1 1] [1 -1] [-1 1] [-1 -1]]]
    (->> directions
         (filter #(xmas-in-direction? pos % input))
         (count))))


(defn total-xmas-count
  [input]
  (let [coords (for [x (range (count input))
                     y (range (count input))]
                 [x y])]
    (->> coords
         (map #(xmas-count % input))
         (apply +))))

;; Part 1
(->> real-input
     (total-xmas-count)
     #_(submit-answer 1))

(defn mas-cross?
  [pos input]
  (let [directions [[1 1] [-1 -1] [1 -1] [-1 1]]
        neighbors (->> directions
                       (map #(map + pos %))
                       (map #(char-at % input)))]
    (and
     (= (char-at pos input) \A)
     (= (apply str (sort neighbors)) "MMSS")
     (not= (apply str neighbors) "MMSS")
     (not= (apply str neighbors) "SSMM"))))

(defn total-mas-cross-count
  [input]
  (let [coords (for [x (range (count input))
                     y (range (count input))]
                 [x y])]
    (->> coords
         (filter #(mas-cross? % input))
         (count))))

;; Part 2
(->> real-input
     (total-mas-cross-count)
     #_(submit-answer 2))
